sdks/python/apache_beam/io/gcp/tests/pubsub_matcher.py

```diff
@@ -0,0 +1,108 @@
+# pylint: disable=wrong-import-order, wrong-import-position
+try:
+  from google.cloud import pubsub
+except ImportError:
+  pubsub = None
+# pylint: enable=wrong-import-order, wrong-import-position
+
+DEFAULT_TIMEOUT = 5 * 60
+MAX_MESSAGES_IN_ONE_PULL = 50
+
+
+class PubSubMessageMatcher(BaseMatcher):
+  """Matcher that verifies messages from given subscription.
+
+  This matcher can block the test and keep pulling messages from given
+  subscription until all expected messages are shown or timeout.
+  """
+
+  def __init__(self, project, sub_name, expected_msg, timeout=DEFAULT_TIMEOUT):
+    """Initialize PubSubMessageMatcher object.
+
+    Args:
+      project: A name string of project.
+      sub_name: A name string of subscription which is attached to output.
+      expected_msg: A string list that contains expected message data pulled
+        from the subscription.
+      timeout: Timeout in seconds to wait for all expected messages appears.
+    """
+    if pubsub is None:
+      raise ImportError(
+          'PubSub dependencies are not installed.')
+    if not project:
+      raise ValueError('Invalid project %s.' % project)
+    if not sub_name:
+      raise ValueError('Invalid subscription %s.' % sub_name)
+    if not expected_msg or not isinstance(expected_msg, list):
+      raise ValueError('Invalid expected messages %s.' % expected_msg)
+
+    self.project = project
+    self.sub_name = sub_name
+    self.expected_msg = expected_msg
+    self.timeout = timeout
+    self.messages = None
+
+  def _matches(self, _):
+    if not self.messages:
+      subscription = (pubsub
+                      .Client(project=self.project)
+                      .subscription(self.sub_name))
+      self.messages = self._wait_for_messages(subscription,
+                                              len(self.expected_msg),
+                                              self.timeout)
+    return Counter(self.messages) == Counter(self.expected_msg)
+
+  def _wait_for_messages(self, subscription, expected_num, timeout):
+    """Wait for messages from given subscription."""
+    logging.debug('Start pulling messages from %s', subscription.full_name)
+    total_messages = []
+    start_time = time.time()
+    while time.time() - start_time <= timeout:
+      pulled = subscription.pull(max_messages=MAX_MESSAGES_IN_ONE_PULL)
+      for ack_id, message in pulled:
+        total_messages.append(message.data)
+        subscription.acknowledge([ack_id])
+      if len(total_messages) >= expected_num:
+        return total_messages
+      time.sleep(1)
+
+    raise RuntimeError('Timeout after %d sec. Received %d messages from %s.' %
+                       (timeout, len(total_messages), subscription.full_name))
+
+  def describe_to(self, description):
+    description.append_text(
+        'Expected %d messages.' % len(self.expected_msg))
+
+  def describe_mismatch(self, _, mismatch_description):
+    diff = set(self.expected_msg) - set(self.messages)
```

Comment:
don't you want to use a counter here to? That is what you used for matching. (For example if you are checking for messages x, x and only x appears, it will mismatch (correctly), however describe mismatch will miss the difference.

Reply:
done